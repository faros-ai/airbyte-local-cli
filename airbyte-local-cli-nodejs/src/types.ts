import {Dictionary} from 'ts-essentials';

export interface AirbyteCliContext {
  tmpDir?: string;
}

// Cli options provided by the user
export interface CliOptions {
  // source and destination config
  configFile?: string;
  wizard?: object;

  // source
  srcImage: string;
  srcConfig?: Record<string, any>;
  srcOutputFile?: string;
  srcCheckConnection?: boolean;
  srcOnly?: boolean;
  srcPull?: boolean;

  // destination
  dstImage: string;
  dstConfig?: Record<string, any>;
  dstUseHostNetwork?: boolean;
  dstOnly?: string;
  dstPull?: boolean;

  // general connector settings
  connectionName?: string;
  stateFile?: string;
  fullRefresh?: boolean;
  rawMessages?: boolean;
  keepContainers?: boolean;
  logLevel?: string;

  // logging
  debug?: boolean;

  // deprecated
  checkConnection?: boolean; // use srcCheckConnection instead
  state?: string; // use stateFile instead
}

// Airbyte connector source or destination config
export interface AirbyteConfig {
  image: string;
  config?: object;
  catalog?: AirbyteConfiguredCatalog;
  dockerOptions?: AirbyteConfigDockerOptions;
}
export interface AirbyteConfigDockerOptions {
  maxMemory?: number; // unit: MB
  maxCpus?: number; // unit: CPU
  maxLogSize?: string; // default: 10m (10MB)
  additionalOptions?: any;
}
export enum AirbyteConfigInputType {
  FILE = 'file',
  OPTION = 'option',
}

// Config that are needed for running this Airbyte Cli
export interface FarosConfig {
  src?: AirbyteConfig;
  dst?: AirbyteConfig;
  srcOutputFile: string | undefined; // if srcOnly is true
  srcInputFile: string | undefined; // if dstOnly is true
  srcCheckConnection: boolean;
  dstUseHostNetwork: boolean;
  srcPull: boolean;
  dstPull: boolean;
  connectionName: string | undefined;
  stateFile: string | undefined;
  fullRefresh: boolean;
  rawMessages: boolean;
  keepContainers: boolean;
  logLevel: string;
  debug: boolean;

  // internal use
  dstStreamPrefix?: string;
}

/**
 * Copy types from faros-ai/airtype-connectors
 * https://github.com/faros-ai/airbyte-connectors/blob/main/faros-airbyte-cdk/src/protocol.ts
 */
export enum AirbyteConnectionStatus {
  SUCCEEDED = 'SUCCEEDED',
  FAILED = 'FAILED',
}
export enum AirbyteMessageType {
  CATALOG = 'CATALOG',
  CONNECTION_STATUS = 'CONNECTION_STATUS',
  LOG = 'LOG',
  RECORD = 'RECORD',
  SPEC = 'SPEC',
  STATE = 'STATE',
  TRACE = 'TRACE',
}
export interface AirbyteMessage {
  readonly type: AirbyteMessageType;
}
export declare class AirbyteConnectionStatusMessage implements AirbyteMessage {
  readonly connectionStatus: {
    status: AirbyteConnectionStatus;
    message?: string;
  };
  readonly type: AirbyteMessageType;
  constructor(connectionStatus: {status: AirbyteConnectionStatus; message?: string});
}
export enum SyncMode {
  FULL_REFRESH = 'full_refresh',
  INCREMENTAL = 'incremental',
}
export enum DestinationSyncMode {
  APPEND = 'append',
  OVERWRITE = 'overwrite',
  APPEND_DEDUP = 'append_dedup',
}
export interface AirbyteStream {
  name: string;
  json_schema?: Dictionary<any>;
  supported_sync_modes?: SyncMode[];
  source_defined_cursor?: boolean;
  default_cursor_field?: string[];
  source_defined_primary_key?: string[][];
  namespace?: string;
}
export interface AirbyteCatalog {
  streams: AirbyteStream[];
}
export declare class AirbyteCatalogMessage implements AirbyteMessage {
  readonly catalog: AirbyteCatalog;
  readonly type: AirbyteMessageType;
  constructor(catalog: AirbyteCatalog);
}
export interface AirbyteConfiguredStream {
  stream: AirbyteStream;
  sync_mode: SyncMode;
  cursor_field?: string[];
  destination_sync_mode?: DestinationSyncMode;
  primary_key?: string[][];
  disabled?: boolean;
}
export interface AirbyteConfiguredCatalog {
  streams: AirbyteConfiguredStream[];
}
