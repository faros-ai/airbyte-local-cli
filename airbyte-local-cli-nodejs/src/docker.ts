import {writeFileSync} from 'node:fs';
import {Writable} from 'node:stream';

import Docker from 'dockerode';

import {AirbyteConnectionStatus, AirbyteConnectionStatusMessage, AirbyteMessageType, FarosConfig} from './types';
import {DEFAULT_STATE_FILE, logger, SRC_CATALOG_FILENAME, SRC_CONFIG_FILENAME} from './utils';

// Constants
const DEFAULT_MAX_LOG_SIZE = '10m';

// Create a new Docker instance
let _docker = new Docker();

// For testing purposes
export function setDocker(docker: Docker): void {
  _docker = docker;
}

export async function checkDockerInstalled(): Promise<void> {
  try {
    await _docker.version();
    logger.debug('Docker is installed and running.');
  } catch (error: any) {
    logger.error('Docker is not installed or running.');
    throw error;
  }
}

export async function pullDockerImage(image: string): Promise<void> {
  logger.info(`Pulling docker image: ${image}`);

  try {
    const stream = await _docker.pull(image);
    await new Promise((resolve, reject) => {
      _docker.modem.followProgress(stream, (err, res) => (err ? reject(err) : resolve(res)));
    });
    logger.info(`Docker image pulled: ${image}`);
  } catch (error: any) {
    logger.error(`Failed to pull docker image: ${image}`);
    throw error;
  }
}

/**
 * Spinning up a docker container to check the source connection.
 * `docker run --rm -v "$tempdir:/configs" $src_docker_options "$src_docker_image"
 *      check --config "/configs/$src_config_filename"`
 *
 * Sample output from the docker container:
 * {"connectionStatus":{"status":"SUCCEEDED"},"type":"CONNECTION_STATUS"}
 * {"connectionStatus":{"status":"FAILED","message":"Faros API key was not provided"},"type":"CONNECTION_STATUS"}
 */
export async function checkSrcConnection(tmpDir: string, image: string, srcConfigFile?: string): Promise<void> {
  logger.info('Validating connection to source...');

  if (!image) {
    throw new Error('Source image is missing.');
  }

  try {
    const cfgFile = srcConfigFile ?? SRC_CONFIG_FILENAME;
    const command = ['check', '--config', `/configs/${cfgFile}`];
    const createOptions: Docker.ContainerCreateOptions = {
      HostConfig: {
        Binds: [`${tmpDir}:/configs`],
        AutoRemove: true,
      },
      platform: 'linux/amd64',
    };

    // create a writable stream to capture the output
    let data = '';
    const outputStream = new Writable({
      write(chunk, _encoding, callback) {
        data += chunk.toString();
        callback();
      },
    });

    // docker run
    const res = await _docker.run(image, command, outputStream, createOptions);

    // capture connection status from the output
    let status: AirbyteConnectionStatusMessage | undefined;
    data.split('\n').forEach((line) => {
      if (line.includes(AirbyteMessageType.CONNECTION_STATUS)) {
        status = JSON.parse(line) as AirbyteConnectionStatusMessage;
      }
    });
    if (
      status?.type === AirbyteMessageType.CONNECTION_STATUS &&
      status?.connectionStatus.status === AirbyteConnectionStatus.SUCCEEDED &&
      res[0].StatusCode === 0
    ) {
      logger.info('Source connection is valid.');
    } else {
      throw new Error(status?.connectionStatus.message);
    }
  } catch (error: any) {
    throw new Error(`Failed to validate source connection: ${error.message ?? JSON.stringify(error)}.`);
  }
}

/**
 * Spinning up a docker container to run source airbyte connector.
 * Platform is set to 'linux/amd64' as we only publish airbyte connectors images for linux/amd64.
 *
 * Docker cli command:
 *  docker run --name $src_container_name --init \
 *    -v "$tempdir:/configs" \
 *    $max_memory $max_cpus --log-opt max-size="$max_log_size" \
 *    --env LOG_LEVEL="$log_level" \
 *    $src_docker_options
 *    --cidfile="$tempPrefix-src_cid" \
 *    -a stdout -a stderr \
 *    "$src_docker_image" \
 *    read \
 *    --config "/configs/$src_config_filename" \
 *    --catalog "/configs/$src_catalog_filename" \
 *    --state "/configs/$src_state_filename"
 *
 * @argument command - for testing purposes only
 */
export async function runSrcSync(tmpDir: string, config: FarosConfig, command?: string[]): Promise<string> {
  logger.info('Running source connector...');

  if (!config.src?.image) {
    throw new Error('Source image is missing.');
  }

  try {
    const timestamp = Date.now();
    const srcContainerName = `airbyte-local-src-${timestamp}`;
    const cmd = command ?? [
      'read',
      '--config',
      `/configs/${SRC_CONFIG_FILENAME}`,
      '--catalog',
      `/configs/${SRC_CATALOG_FILENAME}`,
      '--state',
      `/configs/${DEFAULT_STATE_FILE}`,
    ];
    const maxNanoCpus =
      config.src?.dockerOptions?.maxCpus !== undefined ? config.src.dockerOptions.maxCpus * 1e9 : undefined;
    const maxMemory =
      config.src?.dockerOptions?.maxMemory !== undefined ? config.src.dockerOptions.maxMemory * 1024 * 1024 : undefined;
    const createOptions: Docker.ContainerCreateOptions = {
      name: srcContainerName,
      Image: config.src.image,
      Cmd: cmd,
      HostConfig: {
        Binds: [`${tmpDir}:/configs`],
        AutoRemove: true,
        Init: true,
        NanoCpus: maxNanoCpus, // 1e9 nano cpus = 1 cpu
        Memory: maxMemory, // 1024 * 1024 bytes = 1MB
        LogConfig: {
          Type: 'json-file',
          Config: {
            'max-size': config.src?.dockerOptions?.maxLogSize ?? DEFAULT_MAX_LOG_SIZE,
          },
        },
      },
      AttachStdout: true,
      AttachStderr: true,
      Env: [`LOG_LEVEL=${config.logLevel}`],
      platform: 'linux/amd64',
    };

    // Create the Docker container
    const container = await _docker.createContainer(createOptions);

    // Write the container ID to the cidfile
    const cidfilePath = `tmp-${timestamp}-src_cid`;
    writeFileSync(cidfilePath, container.id);

    // create a writable stream to capture the stdout
    let data = '';
    const stdoutStream = new Writable({
      write(chunk, _encoding, callback) {
        data += chunk.toString();
        callback();
      },
    });
    // Attach the stderr to termincal stderr, and stdout to the output stream
    const stream = await container.attach({stream: true, stdout: true, stderr: true});
    container.modem.demuxStream(stream, stdoutStream, process.stderr);

    // Start the container
    await container.start();

    // Wait for the container to finish
    const res = await container.wait();
    logger.debug(res);
    logger.debug(data);

    if (res.StatusCode === 0) {
      logger.info('Source connector ran successfully.');
    } else {
      throw new Error('Failed to run source connector.');
    }

    // return the stdout data
    return data;
  } catch (error: any) {
    throw new Error(`Failed to run source connector: ${error.message ?? JSON.stringify(error)}.`);
  }
}
