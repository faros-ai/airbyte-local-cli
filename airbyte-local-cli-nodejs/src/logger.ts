import pino from 'pino';
import pretty from 'pino-pretty';

// Create a pino logger instance
// export const logger = pino(
//   {
//     level: 'info',
//     transport: undefined as any, // no transport
//   },
//   pino.destination(1), // sync to stdout
// );
export const logger = pino(pretty({colorize: true}));
