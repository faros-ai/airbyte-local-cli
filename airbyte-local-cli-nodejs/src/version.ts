export const CLI_VERSION = '0.0.5';
