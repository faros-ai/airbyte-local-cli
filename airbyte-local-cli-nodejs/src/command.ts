import {Command, Option} from 'commander';

import {AirbyteConfig, AirbyteConfigInputType, CliOptions, FarosConfig} from './types';
import {logger, parseConfigFile, updateLogLevel} from './utils';
import {CLI_VERSION} from './version';

// Command line program
function command() {
  const cmd = new Command()
    .name('airbyte-local')
    .description('Airbyte local CLI')
    .version(CLI_VERSION, '-v, --version', 'Output the current version')
    // Enable this to allow src and dst configs to be passed as options
    .allowUnknownOption(true)
    // Help configiration
    .helpOption('-h, --help', 'Display usage information')
    .showHelpAfterError('(add --help for additional information)')
    .configureHelp({
      helpWidth: 100,
    })
    // Output configuration
    .configureOutput({
      writeErr: (str) => logger.error(str),
      writeOut: (str) => logger.info(str),
    })

    // Options: Airbyte connector config
    .addOption(
      new Option('--config-file <path>', 'Airbyte source and destination connector config Json file path').conflicts([
        'src',
        'dst',
        'wizard',
      ]),
    )
    // TODO: @FAI-13889 Finalize the wizard arugments and implementation
    .addOption(
      new Option('--wizard <src> [dst]', 'Run the Airbyte configuration wizard').conflicts([
        'configFile',
        'src',
        'dst',
      ]),
    )
    .option('--src <image>', 'Airbyte source Docker image')
    .option('--dst <image>', 'Airbyte destination Docker image')
    .option('--src.<key> <value>', 'Add "key": "value" into the source config')
    .option('--dst.<key> <value>', 'Add "key": "value" into the destination config')
    .option('--full-refresh', 'Force full_refresh and overwrite mode. This overrides the mode in provided config file.')
    .option('--state-file <path>', 'Override state file path for incremental sync')

    // Options: Airbyte connector settings
    .option('--no-src-pull', 'Skip pulling Airbyte source image')
    .option('--no-dst-pull', 'Skip pulling Airbyte destination image')
    .option('--src-check-connection', 'Validate the Airbyte source connection')
    .addOption(
      new Option(
        '--src-only',
        `Only run the Airbyte source and write output in stdout. Use '--src-output-file' instead to write to a file`,
      ).conflicts('srcOutputFile'),
    )
    .option('--src-output-file <path>', 'Write source output as a file (requires a destination)')
    .option('--dst-only <file>', 'Use a file for destination input instead of a source')
    .option('--dst-use-host-network', 'Use the host network when running the Airbyte destination')
    .option('--log-level <level>', 'Set level of source and destination loggers', 'info')
    .option('--raw-messages', 'Output raw Airbyte messages')
    .option('--connection-name <name>', 'Connection name used in various places')
    .option('--keep-containers', 'Do not remove source and destination containers after they exit')

    // Options: Cli settings
    .option('--debug', 'Enable debug logging')

    // Additional check
    .action((opts: any) => {
      if (opts.debug) {
        updateLogLevel(true);
        cmd.setOptionValue('logLevel', 'debug');
      }

      // Check if Airbyte config is provided
      if (!opts.configFile && !(opts.src || opts.dst) && !opts.wizard) {
        cmd.error(
          'Configuration options are missing. Please provide one of the following options: ' +
            `'--config-file', '--src' and '--dst', or '--wizard' to configure the Airbyte connector.`,
        );
      }
    });
  return cmd;
}

// Parse the key-value pair of source and destination configurations
function parseSrcAndDstConfig(argv: string[]) {
  const srcConfig: Record<string, string> = {};
  const dstConfig: Record<string, string> = {};

  function handleNestedConfig(obj: Record<string, any>, keys: string[], value: string) {
    let current = obj;
    keys.forEach((key, index) => {
      if (index === keys.length - 1) {
        current[key] = value;
      } else {
        if (!current[key]) {
          current[key] = {};
        }
        current = current[key];
      }
    });
  }

  argv.forEach((arg, index) => {
    if (arg.startsWith('--src.')) {
      const keys = arg.split('.').slice(1);
      const value = argv[index + 1];
      if (keys.length && value && !value.startsWith('--')) {
        handleNestedConfig(srcConfig, keys, value);
      }
    } else if (arg.startsWith('--dst.')) {
      const keys = arg.split('.').slice(1);
      const value = argv[index + 1];
      if (keys.length && value && !value.startsWith('--')) {
        handleNestedConfig(dstConfig, keys, value);
      }
    }
  });
  logger.debug('Source Config:', srcConfig);
  logger.debug('Destination Config:', dstConfig);
  return {srcConfig, dstConfig};
}

// Convert the options to CliOptions type
function convertToCliOptions(options: any): CliOptions {
  return {
    ...options,
    srcImage: options.src,
    dstImage: options.dst,
  } as CliOptions;
}

// Validate the input options
function validateConfigFileInput(config: FarosConfig, inputType: AirbyteConfigInputType): void {
  if (!config.src?.image && !config.srcInputFile) {
    if (inputType === AirbyteConfigInputType.OPTION) {
      throw new Error(`Missing source image. Please use '--src <image>' to provide the source image`);
    } else {
      throw new Error('Missing source image. Please make sure you provide it in the config file.');
    }
  }
  if (!config.dst?.image && !config.srcOutputFile) {
    if (inputType === AirbyteConfigInputType.OPTION) {
      throw new Error(`Missing destination image. Please use '--dst <image>' to provide the destination image`);
    } else {
      throw new Error('Missing destination image. Please make sure you provide it in the config file.');
    }
  }
  if (!config.src?.config || !config.dst?.config) {
    logger.warn('No source or destination config provided. Please make sure this is intended.');
  }
}

// parse the command line arguments
export function parseAndValidateInputs(argv: string[]): FarosConfig {
  // Parse the command line arguments
  const program = command().parse(argv);

  // Check for unknown options
  const unknown = program.parseOptions(argv).unknown;
  unknown.forEach((u) => {
    if (u.startsWith('--') && !u.startsWith('--src.') && !u.startsWith('--dst.')) {
      throw new Error(`Unknown option: ${u}`);
    }
  });

  // Handle the src and dst config options passed as options
  const configKeyValues = parseSrcAndDstConfig(argv);
  program.setOptionValue('srcConfig', configKeyValues.srcConfig);
  program.setOptionValue('dstConfig', configKeyValues.dstConfig);

  // Get the options
  const options = program.opts();
  logger.debug({options}, 'Options');
  const cliOptions = convertToCliOptions(options);
  logger.debug({cliOptions}, 'Cli options');

  // Convert the cli options to FarosConfig
  const farosConfig: FarosConfig = {
    // The default source output file is stdout if `srcOnly` is true
    // Take the non-default value if provided with `srcOutputFile` option
    srcOutputFile: cliOptions.srcOnly ? '/dev/null' : cliOptions.srcOutputFile,
    // Rename the `dstOnly` file path to `srcInputFile`
    srcInputFile: cliOptions.dstOnly,
    connectionName: cliOptions.connectionName,
    stateFile: cliOptions.stateFile,
    srcCheckConnection: cliOptions.srcCheckConnection ?? false,
    dstUseHostNetwork: cliOptions.dstUseHostNetwork ?? false,
    // If `dstOnly` is true, do not pull the source image. Otherwise, fall back to option `noSrcPull`
    srcPull: (cliOptions.dstOnly ? false : cliOptions.srcPull) ?? true,
    dstPull: (cliOptions.srcOnly ? false : cliOptions.dstPull) ?? true,
    fullRefresh: cliOptions.fullRefresh ?? false,
    rawMessages: cliOptions.rawMessages ?? false,
    keepContainers: cliOptions.keepContainers ?? false,
    logLevel: cliOptions.logLevel ?? 'info',
    debug: cliOptions.debug ?? false,
  };

  if (cliOptions.srcImage || cliOptions.dstImage) {
    farosConfig.src = {
      image: cliOptions.srcImage,
      config: cliOptions.srcConfig,
    } as AirbyteConfig;
    farosConfig.dst = {
      image: cliOptions.dstImage,
      config: cliOptions.dstConfig,
    } as AirbyteConfig;
    validateConfigFileInput(farosConfig, AirbyteConfigInputType.OPTION);
  } else if (cliOptions.configFile) {
    logger.info('Reading config file...');
    const airbyteConfig = parseConfigFile(cliOptions.configFile);
    farosConfig.src = airbyteConfig.src;
    farosConfig.dst = airbyteConfig.dst;
    validateConfigFileInput(farosConfig, AirbyteConfigInputType.FILE);
  } else if (cliOptions.wizard) {
    logger.info('Runing wizard...');
    // TODO: @FAI-13889 Implement the wizard
  }
  return farosConfig;
}
