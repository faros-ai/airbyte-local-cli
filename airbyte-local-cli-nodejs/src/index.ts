import {parseAndValidateInputs} from './command';
import {checkDockerInstalled, checkSrcConnection, pullDockerImage, runSrcSync} from './docker';
import {AirbyteCliContext} from './types';
import {cleanUp, createTmpDir, loadStateFile, logger, writeConfig} from './utils';

async function main(): Promise<void> {
  const context: AirbyteCliContext = {};
  try {
    // Parse and validate cli arguments
    const cfg = parseAndValidateInputs(process.argv);
    await checkDockerInstalled();

    // Create temporary directory, load state file, write config to files
    context.tmpDir = createTmpDir();
    loadStateFile(context.tmpDir, cfg?.stateFile, cfg?.connectionName);
    writeConfig(context.tmpDir, cfg);

    // Pull source docker image
    if (cfg.srcPull && cfg.src?.image) {
      await pullDockerImage(cfg.src.image);
    }
    // Check source connection
    if (cfg.srcCheckConnection && cfg.src?.image) {
      await checkSrcConnection(context.tmpDir, cfg.src.image);
    }
    // Run airbyte source connector
    if (!cfg.srcInputFile) {
      await runSrcSync(context.tmpDir, cfg);
    }
  } catch (error: any) {
    logger.error(error.message, 'Error');
    cleanUp(context);
    throw error;
  }
}

main().catch((_error) => {
  logger.error('Exit Airbyte CLI with errors.');
  process.exit(1);
});
