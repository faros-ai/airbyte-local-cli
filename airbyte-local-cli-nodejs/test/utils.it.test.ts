import {existsSync, mkdtempSync, readFileSync, rmSync} from 'node:fs';
import {tmpdir} from 'node:os';

import {FarosConfig} from '../src/types';
import {
  checkDockerInstalled,
  cleanUp,
  createTmpDir,
  FILENAME_PREFIX,
  loadStateFile,
  parseConfigFile,
  writeConfig,
} from '../src/utils';

describe('parseConfigFile', () => {
  it('should pass', () => {
    expect(parseConfigFile('test/resources/test_config_file.json')).toMatchSnapshot();
  });
  it('should fail with invalid json', () => {
    expect(() => parseConfigFile('test_config_file_invalid')).toThrow();
  });
});

describe('checkDockerInstalled', () => {
  it('should pass if docker is installed', () => {
    expect(checkDockerInstalled('pwd', [])).toBeUndefined();
  });

  it('should fail if docker is not installed', () => {
    expect(() => checkDockerInstalled('bad-command')).toThrow();
  });
});

describe('createTmpDir', () => {
  it('should pass', () => {
    const tmpDirPath = createTmpDir();
    expect(existsSync(tmpDirPath)).toBe(true);
    expect(() => cleanUp({tmpDir: tmpDirPath})).not.toThrow();
  });

  it('should pass with non default dir', () => {
    const tmpDirPath = createTmpDir(`test/test-tmp-dir`);
    expect(existsSync(tmpDirPath)).toBe(true);
    expect(() => cleanUp({tmpDir: tmpDirPath})).not.toThrow();
  });
});

describe('write files to temporary dir', () => {
  let tmpDirPath: string;
  beforeAll(() => {
    tmpDirPath = mkdtempSync(`${tmpdir()}/test-temp-dir`);
  });
  afterAll(() => {
    rmSync(tmpDirPath, {recursive: true, force: true});
  });

  describe('loadStateFile', () => {
    it('should pass without existing state file', () => {
      expect(() => loadStateFile(tmpDirPath)).not.toThrow();
      expect(existsSync(`${tmpDirPath}/state.json`)).toBe(true);
      expect(readFileSync(`${tmpDirPath}/state.json`, 'utf8')).toBe('{}');
    });

    it('should pass with existing state file', () => {
      const testStateFile = 'test/resources/test__state.json';
      expect(() => loadStateFile(tmpDirPath, testStateFile)).not.toThrow();
      expect(existsSync(`${tmpDirPath}/state.json`)).toBe(true);
      expect(readFileSync(`${tmpDirPath}/state.json`, 'utf8')).toMatchSnapshot();
    });

    it('should fail if state file is not loaded', () => {
      expect(() => loadStateFile(tmpDirPath, 'non-exist-state-file')).toThrow(
        `State file 'non-exist-state-file' not found. Please make sure the state file exists and have read access.`,
      );
    });
  });

  describe('writeConfig', () => {
    const testConfig: FarosConfig = {
      src: {
        image: 'farosai/airbyte-test-source',
        config: {
          username: 'test',
          password: 'test',
          url: 'test',
        },
        catalog: {
          tests: {disabled: true},
          projects: {disabled: true},
        },
      },
      dst: {
        image: 'farosai/airbyte-test-destination',
        config: {
          edition_config: {
            graph: 'default',
            edition: 'cloud',
            api_url: 'https://test.api.faros.ai',
          },
        },
      },

      // default values
      srcCheckConnection: false,
      dstUseHostNetwork: false,
      srcPull: false,
      dstPull: false,
      fullRefresh: false,
      rawMessages: false,
      keepContainers: false,
      logLevel: 'info',
      debug: false,
      stateFile: undefined,
      connectionName: undefined,
      srcOutputFile: undefined,
      srcInputFile: undefined,
    };

    afterEach(() => {
      rmSync(`${FILENAME_PREFIX}_config.json`, {force: true});
      rmSync(`${tmpDirPath}/${FILENAME_PREFIX}_src_config.json`, {force: true});
      rmSync(`${tmpDirPath}/${FILENAME_PREFIX}_dst_config.json`, {force: true});
      rmSync(`${tmpDirPath}/${FILENAME_PREFIX}_src_catalog.json`, {force: true});
      rmSync(`${tmpDirPath}/${FILENAME_PREFIX}_dst_catalog.json`, {force: true});
    });

    it('should write files', () => {
      expect(() => writeConfig(tmpDirPath, structuredClone(testConfig))).not.toThrow();
      expect(existsSync(`${FILENAME_PREFIX}_config.json`)).toBe(true);
      expect(existsSync(`${tmpDirPath}/${FILENAME_PREFIX}_src_config.json`)).toBe(true);
      expect(existsSync(`${tmpDirPath}/${FILENAME_PREFIX}_dst_config.json`)).toBe(true);
      expect(existsSync(`${tmpDirPath}/${FILENAME_PREFIX}_src_catalog.json`)).toBe(true);
      expect(existsSync(`${tmpDirPath}/${FILENAME_PREFIX}_dst_catalog.json`)).toBe(true);

      expect(readFileSync(`${FILENAME_PREFIX}_config.json`, 'utf8')).toEqual(
        JSON.stringify({src: testConfig.src, dst: testConfig.dst}, null, 2),
      );
      expect(readFileSync(`${tmpDirPath}/${FILENAME_PREFIX}_src_config.json`, 'utf8')).toEqual(
        JSON.stringify(testConfig.src?.config, null, 2),
      );
      expect(readFileSync(`${tmpDirPath}/${FILENAME_PREFIX}_dst_config.json`, 'utf8')).toEqual(
        JSON.stringify(testConfig.dst?.config, null, 2),
      );
      expect(readFileSync(`${tmpDirPath}/${FILENAME_PREFIX}_src_catalog.json`, 'utf8')).toEqual(
        JSON.stringify(testConfig.src?.catalog, null, 2),
      );
      expect(readFileSync(`${tmpDirPath}/${FILENAME_PREFIX}_dst_catalog.json`, 'utf8')).toEqual(
        JSON.stringify(testConfig.src?.catalog, null, 2),
      );
    });

    it('should alter config if debug is enabled', () => {
      const testConfigDebug = {...structuredClone(testConfig), debug: true};
      (testConfigDebug.src as any).image = 'farosai/airbyte-faros-feeds-source:v1';
      expect(() => writeConfig(tmpDirPath, structuredClone(testConfigDebug))).not.toThrow();
      expect(existsSync(`${FILENAME_PREFIX}_config.json`)).toBe(true);
      expect(existsSync(`${tmpDirPath}/${FILENAME_PREFIX}_src_config.json`)).toBe(true);
      expect(existsSync(`${tmpDirPath}/${FILENAME_PREFIX}_dst_config.json`)).toBe(true);

      expect(readFileSync(`${FILENAME_PREFIX}_config.json`, 'utf8')).toEqual(
        JSON.stringify({src: testConfigDebug.src, dst: testConfigDebug.dst}, null, 2),
      );
      expect(readFileSync(`${tmpDirPath}/${FILENAME_PREFIX}_src_config.json`, 'utf8')).toEqual(
        JSON.stringify({...testConfigDebug.src?.config, feed_cfg: {debug: true}}, null, 2),
      );
      expect(readFileSync(`${tmpDirPath}/${FILENAME_PREFIX}_dst_config.json`, 'utf8')).toEqual(
        JSON.stringify(testConfigDebug.dst?.config, null, 2),
      );
    });
  });
});
