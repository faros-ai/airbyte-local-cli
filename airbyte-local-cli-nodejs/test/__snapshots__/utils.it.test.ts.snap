// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseConfigFile should pass 1`] = `
{
  "dst": {
    "catalog": {},
    "config": {
      "edition_config": {
        "api_key": "***",
        "api_url": "https://prod.api.faros.ai",
        "edition": "cloud",
        "graph": "default",
      },
    },
    "dockerOptions": {
      "maxCpus": 2,
      "maxMemory": 2048,
    },
    "image": "farosai/airbyte-servicenow-destination",
  },
  "src": {
    "catalog": {},
    "config": {
      "password": "***",
      "url": "https://test-instance.service-now.com",
      "username": "test-username",
    },
    "dockerOptions": {
      "maxCpus": 2,
      "maxMemory": 2048,
    },
    "image": "farosai/airbyte-servicenow-source",
  },
}
`;

exports[`processSrcInputFile should succeed writing to an output file 1`] = `
"{"state":{"data":{"format":"base64/gzip","data":"H4sIAAAAAAAAA6uuBQBDv6ajAgAAAA=="}},"type":"STATE","redactedConfig":{"user":"chris"},"sourceType":"example","sourceVersion":"0.12.3"}
{"state":{"data":{"format":"base64/gzip","data":"H4sIAAAAAAAAA6uuBQBDv6ajAgAAAA=="}},"type":"STATE","sourceStatus":{"status":"SUCCESS"}}
{"state":{"data":{"format":"base64/gzip","data":"H4sIAAAAAAAAA6uuBQBDv6ajAgAAAA=="}},"type":"STATE","logs":[{"timestamp":***,"message":{"level":30,"msg":"Source version: 0.12.3"}},{"timestamp":***,"message":{"level":30,"msg":"Config: {\\"user\\":\\"chris\\"}"}},{"timestamp":***,"message":{"level":30,"msg":"Catalog: {}"}},{"timestamp":***,"message":{"level":30,"msg":"State: {}"}},{"timestamp":***,"message":{"level":30,"msg":"Syncing ExampleSource"}},{"timestamp":***,"message":{"level":30,"msg":"Finished syncing ExampleSource"}}]}
"
`;

exports[`write files to temporary dir loadStateFile should pass with existing state file 1`] = `
"{"format":"base64/gzip","data":"dGVzdA=="}
"
`;

exports[`write files to temporary dir writeCatalog should succeed with default only 1`] = `
{
  "streams": [
    {
      "destination_sync_mode": "append",
      "stream": {
        "json_schema": {},
        "name": "builds",
        "supported_sync_modes": [
          "full_refresh",
          "incremental",
        ],
      },
      "sync_mode": "incremental",
    },
  ],
}
`;

exports[`write files to temporary dir writeCatalog should succeed with default only 2`] = `
{
  "streams": [
    {
      "destination_sync_mode": "append",
      "stream": {
        "json_schema": {},
        "name": "testPrefix__builds",
        "supported_sync_modes": [
          "full_refresh",
          "incremental",
        ],
      },
      "sync_mode": "incremental",
    },
  ],
}
`;

exports[`write files to temporary dir writeCatalog should succeed with dst override 1`] = `
{
  "streams": [
    {
      "destination_sync_mode": "append",
      "stream": {
        "json_schema": {},
        "name": "builds",
        "supported_sync_modes": [
          "full_refresh",
          "incremental",
        ],
      },
      "sync_mode": "incremental",
    },
  ],
}
`;

exports[`write files to temporary dir writeCatalog should succeed with dst override 2`] = `
{
  "streams": [
    {
      "destination_sync_mode": "overwrite",
      "stream": {
        "json_schema": {},
        "name": "testPrefix__builds",
        "supported_sync_modes": [
          "full_refresh",
          "incremental",
        ],
      },
      "sync_mode": "full_refresh",
    },
  ],
}
`;

exports[`write files to temporary dir writeCatalog should succeed with src override 1`] = `
{
  "streams": [
    {
      "destination_sync_mode": "overwrite",
      "stream": {
        "json_schema": {},
        "name": "builds",
        "supported_sync_modes": [
          "full_refresh",
          "incremental",
        ],
      },
      "sync_mode": "full_refresh",
    },
  ],
}
`;

exports[`write files to temporary dir writeCatalog should succeed with src override 2`] = `
{
  "streams": [
    {
      "destination_sync_mode": "overwrite",
      "stream": {
        "json_schema": {},
        "name": "testOverridePrefix__builds",
        "supported_sync_modes": [
          "full_refresh",
          "incremental",
        ],
      },
      "sync_mode": "full_refresh",
    },
  ],
}
`;
